Modernized RawPCCProducer

Dr15Jones opened the pull request:
The order in which endLuminosityBlock and endLuminosityBlockProduce
are called by the framework was reversed to match the behavior in
global modules. This required moving the computation from one function
to the other function.

In addition
-many member data were converted to local variables
-member data were marked as 'const' if possible
-switched to using edm::EDPutTokenT
These changes would make it easy to convert this module to a global
module in the future. This was not done now because of the output file.

Dr15Jones committed:
Modernized RawPCCProducer

The order in which endLuminosityBlock and endLuminosityBlockProduce
are called by the framework was reversed to match the behavior in
global modules. This required moving the computation from one function
to the other function.

In addition
-many member data were converted to local variables
-member data were marked as 'const' if possible
-switched to using edm::EDPutTokenT
These changes would make it easy to convert this module to a global
module in the future. This was not done now because of the output file.

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-25263/7284

cmsbuild commented:
A new Pull Request was created by  @Dr15Jones (Chris Jones) for master.

It involves the following packages:

Calibration/LumiAlCaRecoProducers


@cmsbuild, @franzoni, @pohsun, @tocheng, @lpernie can you please review it and eventually sign? Thanks.
@mmusich, @tocheng this is something you requested to watch as well.
@davidlange6, @slava77, @fabiocos you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

Dr15Jones commented:
NOTE: I was hoping to convert RawPCCProducer into a `edm::global` module in order to support concurrent LuminosityBlock processing. I got very close but I was stopped because the module optionally writes out a text file. At begin LuminosityBlock it opens the file and starts a new line, then at end LuminosityBlock it finishes writing that line. Clearly, such semantics does not support concurrent LuminosityBlock processing. There are several possibilities

- remove the option to create the text file
- do not write during begin LuminosityBlock and instead write the entire line at end LuminosityBlock
- write two lines per LuminosityBlock, one line at begin and another at end

The choice on which of these options is the best depends on how the text file is used.

Dr15Jones commented:
please test workflow 1020.0

cmsbuild commented:
The tests are being triggered in jenkins.

Dr15Jones commented on Calibration/LumiAlCaRecoProducers/plugins/RawPCCProducer.cc:
```diff
@@ -43,55 +43,33 @@ class RawPCCProducer : public edm::one::EDProducer<edm::EndLuminosityBlockProduc
-    produces<LumiInfo, edm::Transition::EndLuminosityBlock>(outputProductName_);
-
-    if(!applyCorr_){
-        correctionScaleFactors_.resize(LumiConstants::numBX,1.0);
```
Setting of the size of the container now happens at line 146

Dr15Jones commented on Calibration/LumiAlCaRecoProducers/plugins/RawPCCProducer.cc:
```diff
@@ -107,113 +85,122 @@ void RawPCCProducer::produce(edm::Event& iEvent, const edm::EventSetup& iSetup){
-    statErrOnLumi_=0.0;
+}
 
-    std::fill(clustersPerBXOutput_.begin(),  clustersPerBXOutput_.end(),0);
```
These fills are no longer necessary since I create new containers each `endLuminosityBlockProduce` call and they are constructed with the correct size and filled with 0.

Dr15Jones commented:
Many of the changes shown on a line are because when I changed data members to just be local variables I also removed the trailing underscore from their names to conform to CMS naming conventions.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-25263/31696/summary.html

@slava77 comparisons for the following workflows were not done due to missing matrix map:
   * /build/cmsbld/jenkins/workspace/compare-root-files-short-matrix/results/JR-comparison/PR-25263/1020.0_AlCaLumiPixels2016H+AlCaLumiPixels2016H+TIER0EXPLP+ALCAEXPLP+ALCAHARVLP+TIER0PROMPTLP

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 1072 differences found in the comparisons
 - DQMHistoTests: Total files compared: 32
 - DQMHistoTests: Total histograms compared: 3013311
 - DQMHistoTests: Total failures: 58
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3013056
 - DQMHistoTests: Total skipped: 197
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 31 files compared)
 - Checked 134 log files, 14 edm output root files, 32 DQM output files

fabiocos commented:
@pohsun, @tocheng, @lpernie this PR fixes a problem in the IB, and updates the code. Could you please check it and sign it in case, or comment?

cmsbuild commented:
+1
Tested at: a819374a545ed591132d86da4e91080da2bd55be
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-25263/31720/summary.html

cmsbuild commented:
Comparison job queued.

fabiocos commented:
@pohsun, @tocheng, @lpernie this PR fixes a problem in the IB, and updates the code. Could you please check it and sign it in case, or comment?

fabiocos commented:
No difference is observed in the DQM output of workflow 1020

pohsun commented:
+1

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @davidlange6, @slava77, @smuzaffar, @fabiocos (and backports should be raised in the release meeting by the corresponding L2)

fabiocos commented:
+1

the residual DQM changes look unrelated to this PR
